### PR TITLE
feat(gen-docs): markdown docs for rules

### PR DESCRIPTION
**Pull request by KSXGitHub**
Restructures `gen-docs` into three subcommands sharing the existing extractor:

- `html <out_dir>` — the existing self-contained catalogue page.
- `write-md <rules_dir>` — one markdown file per rule plus a `README.md` index, written under `rules/`.
- `check-md <rules_dir>` — re-renders in memory and fails on missing / stale / orphan files.

`just doc` runs `check-md`, so CI catches any drift between `src/rules/` and the in-tree markdown copy. `rules/` is seeded from the 14 currently implemented rules.

### CODE REVIEW

File: rules/derive_ordering.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+```rust,ignore
+#[derive(Clone, Copy, Debug)]
+struct Point;
+```
+
+## Configuration
+
+Configure via `dylint.toml` under `["perfectionist::derive_ordering"]`. Every field is optional; the per-field prose below states the default.
+
+### `style` — `Style` (optional)
```
Comment (KSXGitHub) [inline rules/derive_ordering.md]: Are you sure em dash can serve this function?

File: rules/README.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+| [`prefer_raw_string`](./prefer_raw_string.md) | `warn` | string literal contains only raw-expressible escapes; prefer the raw-string form |
+| [`single_letter_closure_param`](./single_letter_closure_param.md) | `warn` | closure parameter has a single-letter name |
+| [`single_letter_function_param`](./single_letter_function_param.md) | `warn` | function parameter has a single-letter name |
+| [`single_letter_generic`](./single_letter_generic.md) | `warn` | generic type parameter has a single-letter name |
+| [`single_letter_let_binding`](./single_letter_let_binding.md) | `warn` | `let` binding has a single-letter name |
+| [`unicode_ellipsis_in_comments`](./unicode_ellipsis_in_comments.md) | `warn` | U+2026 HORIZONTAL ELLIPSIS in non-doc comments; prefer `...` |
+| [`unicode_ellipsis_in_panic_messages`](./unicode_ellipsis_in_panic_messages.md) | `warn` | U+2026 HORIZONTAL ELLIPSIS in panic / assertion / expect messages; prefer `...` |
+| [`unknown_perfectionist_lints`](./unknown_perfectionist_lints.md) | `warn` | lint-control attribute references a `perfectionist::*` lint that this plugin does not register |
```
Comment (KSXGitHub) [inline rules/README.md]: This looks ugly in markdown code form.

Maybe switch to a bullet list?

File: rules/README.md
Comment (KSXGitHub) [inline rules/README.md:1]: Also, unrelated to this file,

But some of `planned-rules/` reference already implemented rules.

You may switch them to reference the files in `rules/`

Comment (KSXGitHub) [inline rules/README.md:1]: Add (generate) HTML comments `<!-- ... -->` at the top of every markdown file, telling the reader of the code that these files are generated, not manually written.

File: tools/gen-docs/src/check_md.rs [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+        Err(error) if error.kind() == std::io::ErrorKind::NotFound => return names,
+        Err(error) => panic!("failed to read {}: {error}", rules_dir.display()),
+    };
+    for entry in entries {
+        let entry = entry.expect("failed to read directory entry");
+        let file_type = entry.file_type().expect("failed to read file type");
+        if !file_type.is_file() {
+            continue;
+        }
+        let name = entry.file_name().to_string_lossy().into_owned();
```
Comment (KSXGitHub) [inline tools/gen-docs/src/check_md.rs]: **Finding 1 (minor): non-UTF-8 filename coerced**

`to_string_lossy()` mangles a hypothetical `foo\xFF.md` entry into `foo\u{FFFD}.md`, so the subsequent `rules_dir.join(name)` resolves to the wrong path — orphan-delete and stale-detection would both misbehave.

**Assessment:** Valid but low-probability — every entry under `rules/` is named after a rule which is `snake_case` ASCII. Still cheap to guard. **Will fix:** switch to `entry.file_name().to_str()` and skip entries that aren't valid UTF-8.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tools/gen-docs/src/check_md.rs [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+        let path = rules_dir.join(file_name);
+        // Only write when the content differs, so unchanged files
+        // keep their original mtime and don't churn the working tree
+        // on every `write-md` run.
+        let needs_write = match fs::read_to_string(&path) {
+            Ok(existing) => &existing != body,
+            Err(_) => true,
+        };
+        if needs_write {
+            fs::write(&path, body)
```
Comment (KSXGitHub) [inline tools/gen-docs/src/check_md.rs:169]: **Finding 2 (minor): non-atomic writes**

`fs::write` is not atomic; a process kill mid-write leaves a truncated file that the next `check-md` reports as drift.

**Assessment:** Valid but the operation is short-lived and the recovery action is just to re-run `just gen-rules-md`. The HTML renderer (`run_html`) uses the same non-atomic `fs::write` for `index.html`, so adopting tempfile+rename here only would be asymmetric. **Won't fix** — out of scope; track separately if it ever bites in practice.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tools/gen-docs/src/check_md.rs [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+    } else {
+        let mut full = String::from("Markdown catalogue is out of date:\n");
+        full.push_str(&report);
+        CheckOutcome::Drifted(full)
+    }
+}
+
+pub(crate) fn write_rules_dir(rules: &[Rule], rules_dir: &Path) -> WriteSummary {
+    let expected = ExpectedTree::from_rules(rules);
+    fs::create_dir_all(rules_dir).expect("failed to create rules directory");
```
Comment (KSXGitHub) [inline tools/gen-docs/src/check_md.rs:134]: **Finding 3 (minor): panics on every filesystem error**

`expect(...)` / `panic!(...)` for create-dir / remove / write errors produces an ugly backtrace on routine "permission denied" or "read-only filesystem". The check side returns a structured `CheckOutcome`; the write side could mirror it.

**Assessment:** Valid but matches the panic style already used in `run_html` for `fs::create_dir_all`, `fs::write`, and the manifest parse. Fixing only the write side breaks the symmetry; fixing both is a separate refactor. **Won't fix** — out of scope.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tools/gen-docs/src/main.rs [hunk 6]
```diff
@@ -131,3 +180,54 @@ fn main() -> ExitCode {
...
+fn run_write_md(root: &Path, rules_dir: &Path) -> ExitCode {
+    let rules = match load_rules(root) {
+        Ok(rules) => rules,
+        Err(code) => return code,
+    };
+    let summary = write_rules_dir(&rules, rules_dir);
+    eprintln!(
+        "wrote {} rule file(s) and {} index to {} ({} orphan(s) removed)",
+        summary.rules_written,
+        if summary.index_written { "1" } else { "no" },
```
Comment (KSXGitHub) [inline tools/gen-docs/src/main.rs]: **Finding 4 (nit): summary log reads as if the index was skipped**

```
wrote 0 rule file(s) and no index to ./rules (0 orphan(s) removed)
```

On an idempotent re-run this implies the index wasn't emitted at all, when really it was rendered and matched on-disk content.

**Assessment:** Valid. **Will fix:** rephrase the summary so the unchanged case reads as "unchanged" rather than "no", and rename the field for accuracy.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tools/gen-docs/src/check_md.rs [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+            relative_source: PathBuf::from(format!("src/rules/{name}.rs")),
+            config: None,
+        }
+    }
+
+    /// Build a temporary directory under `target/` so the test
+    /// doesn't litter the working tree, and doesn't conflict with
+    /// another test running in parallel.
+    fn tempdir(label: &str) -> PathBuf {
+        let base = std::env::temp_dir().join(format!(
```
Comment (KSXGitHub) [inline tools/gen-docs/src/check_md.rs:262]: **Finding 5 (nit): tempdir naming relies on per-test label uniqueness**

`format!("perfectionist-gen-docs-{label}-{}", std::process::id())` is unique per process but two tests sharing a `label` inside the same `cargo test` binary would collide and cause a flake.

**Assessment:** Valid in theory; every current call site uses a distinct label so it's not biting today. **Will fix:** add a `static AtomicU64` counter to the path so the invariant is mechanical rather than convention-based.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tools/gen-docs/src/render_md.rs [hunk 8]
```diff
@@ -0,0 +1,375 @@
...
+        }
+    }
+    out.push('\n');
+}
+
+fn unnamespaced(namespaced: &str) -> &str {
+    namespaced.strip_prefix(NAMESPACE).unwrap_or(namespaced)
+}
+
+fn source_path_str(rule: &Rule) -> String {
```
Comment (KSXGitHub) [inline tools/gen-docs/src/render_md.rs:284]: **Finding 6 (nit): `source_path_str` is a roundabout component join**

Iterating `.components()` and lossy-converting each is equivalent to `to_string_lossy().replace(MAIN_SEPARATOR, "/")` and harder to read at a glance.

**Assessment:** Valid but the component form makes the cross-platform intent explicit. **Will fix:** add a one-line comment explaining the `/`-normalisation intent (cheaper than restructuring; preserves the existing logic verbatim).

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by KSXGitHub — COMMENTED** [review]
Self-review (iteration 1) on commit `57a1be5`. Six findings: three I'll fix (UTF-8 guard, summary wording, tempdir counter + path-norm comment), three I'll leave with rationale (non-atomic writes and panic-on-IO match the existing `run_html` style; the cross-ref cleanup the maintainer flagged earlier is still pending as a follow-up). No critical issues, good test coverage, project conventions followed.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: rules/README.md [hunk 9]
```diff
@@ -0,0 +1,22 @@
+<!-- Generated by `gen-docs write-md` from `src/rules/`. Do not edit by hand; run `just gen-rules-md` after changing a rule source. -->
+
+# perfectionist lints
+
+Auto-generated catalogue of implemented lints. Each entry links to a per-rule markdown file rendered from the same source as the [HTML catalogue](https://ksxgithub.github.io/perfectionist/). Edit the rule source in `src/rules/`, not these files; the in-tree copy is regenerated by `just gen-rules-md`.
+
+Lint-control attributes use the `perfectionist::` namespace.
+
+- [`arc_rc_clone`](./arc_rc_clone.md) (default: `warn`). calling `.clone()` on an `Arc<T>` or `Rc<T>`; prefer the qualified `Arc::clone` / `Rc::clone` form
```
Comment (KSXGitHub) [inline rules/README.md]: Maybe multiline?


```suggestion
- [`arc_rc_clone`](./arc_rc_clone.md) (default: `warn`).

  calling `.clone()` on an `Arc<T>` or `Rc<T>`; prefer the qualified `Arc::clone` / `Rc::clone` form

```

File: tools/gen-docs/src/render_md.rs [hunk 10]
```diff
@@ -0,0 +1,394 @@
...
+    let _ = writeln!(
+        out,
+        "**Default level:** `{}`  ",
+        rule.level.to_string().to_ascii_lowercase()
+    );
+    // The per-rule markdown lives one level down from the repo
+    // root (typically `rules/<name>.md`), so the link to the rule
+    // source has to climb back up before descending into `src/`.
+    let source_path = source_path_str(rule);
+    let _ = writeln!(out, "**Source:** [`{source_path}`](../{source_path})");
```
Comment (KSXGitHub) [inline tools/gen-docs/src/render_md.rs]: **Finding 1 (minor): hard-coded `../` source-link prefix**

If `rules_dir` ever lives more than one level below the repo root (e.g. `docs/lints/rules/`), the link `../src/rules/<rule>.rs` climbs one level too few. Today every invocation (`just gen-rules-md` and the manual `gen-docs write-md ./rules`) puts `rules_dir` exactly one level down, but the renderer encodes that as a contract without naming it.

**Assessment:** Valid. **Will fix:** compute the prefix from `rules_dir` relative to `root` and thread it through to `render_rule_md`. Removes a latent shoot-yourself-in-the-foot edge.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tools/gen-docs/src/check_md.rs [hunk 11]
```diff
@@ -0,0 +1,314 @@
...
+                // already guarantees rule namespaces are unique, so
+                // colliding filenames would be a bug in
+                // `rule_file_name`.
+                panic!(
+                    "internal error: two rules rendered to the same filename `{}`",
+                    rule_file_name(rule),
+                );
+            }
+        }
+        files.insert(INDEX_FILE_NAME.to_owned(), render_index_md(rules));
```
Comment (KSXGitHub) [inline tools/gen-docs/src/check_md.rs:45]: **Finding 2 (minor): collision risk when `rules_dir` resolves to the repo root**

If a user runs `gen-docs write-md .` from the repo root, `INDEX_FILE_NAME == "README.md"` collides with the project's own `README.md`. The orphan check considers it managed, so it'd be overwritten silently.

**Assessment:** Valid edge case. **Will fix:** refuse with a clear error when `rules_dir.canonicalize() == root.canonicalize()`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tools/gen-docs/src/check_md.rs [hunk 12]
```diff
@@ -0,0 +1,314 @@
...
+        ExpectedTree { files }
+    }
+
+    fn is_managed(&self, file_name: &str) -> bool {
```
Comment (KSXGitHub) [inline tools/gen-docs/src/check_md.rs:56]: **Finding 3 (minor): case-insensitive filesystems can flag valid files as orphans**

On macOS APFS (case-insensitive by default), a manually-created `readme.md` round-trips through `read_dir` with that casing, fails to match expected `README.md`, and gets deleted as an orphan on the next `write-md`.

**Assessment:** Valid but unreachable in the supported workflow — every file in `rules/` is either generated by this tool (consistent casing) or non-`.md` (ignored). A user manually creating `readme.md` in the managed directory and expecting it to survive is well outside the design. **Won't fix:** add a comment noting the case-sensitivity assumption.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tools/gen-docs/src/main.rs [hunk 13]
```diff
@@ -131,3 +180,58 @@ fn main() -> ExitCode {
...
+                "{} is up to date ({} rule(s))",
+                rules_dir.display(),
+                rules.len(),
+            );
+            ExitCode::SUCCESS
+        }
+        CheckOutcome::Drifted(report) => {
+            eprint!("{report}");
+            eprintln!(
+                "{} is out of date. Run `gen-docs write-md {}` to regenerate.",
```
Comment (KSXGitHub) [inline tools/gen-docs/src/main.rs]: **Finding 4 (nit): error message omits the canonical regeneration entry point**

A user reading the failure should be pointed at `just gen-rules-md` (the wired-up entry point) rather than the bare CLI form, which would have to be parameterised with `--root` to be correct in a non-default workspace.

**Assessment:** Valid. **Will fix:** swap to `just gen-rules-md`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tools/gen-docs/src/check_md.rs [hunk 14]
```diff
@@ -0,0 +1,314 @@
...
+                report.push_str(&format!("- `{}`: missing\n", file_name));
+            }
+            Err(error) => {
+                report.push_str(&format!("- `{}`: read error: {error}\n", file_name));
+            }
+        }
+    }
+
+    for file_name in &actual_names {
+        if !expected.is_managed(file_name) {
```
Comment (KSXGitHub) [inline tools/gen-docs/src/check_md.rs:114]: **Finding 5 (nit): missing test for nonexistent / empty `rules_dir`**

`read_managed_filenames` returns an empty set on `NotFound` — sensible — and `check_rules_dir` then reports every expected file as missing. Reasonable behaviour, not covered by an explicit test.

**Assessment:** Valid. **Will fix:** add a test that runs `check_rules_dir` against a never-created directory and asserts the drift report lists every expected file as missing.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tools/gen-docs/src/check_md.rs [hunk 15]
```diff
@@ -0,0 +1,314 @@
...
+        let needs_write = match fs::read_to_string(&path) {
+            Ok(existing) => &existing != body,
+            Err(_) => true,
+        };
+        if needs_write {
+            fs::write(&path, body)
+                .unwrap_or_else(|error| panic!("failed to write {}: {error}", path.display()));
+        }
+        if file_name == INDEX_FILE_NAME {
+            summary.index_changed = needs_write;
```
Comment (KSXGitHub) [inline tools/gen-docs/src/check_md.rs]: **Finding 6 (nit): `index_changed = needs_write` is plain assignment, not OR**

If the loop ever processed `INDEX_FILE_NAME` more than once (multiple index entries, hypothetically), the last write would silently win.

**Assessment:** Valid but unreachable — `ExpectedTree::from_rules` only inserts the index once, and the loop iterates a `BTreeMap` keyed by filename so duplicates are impossible by construction. **Won't fix:** add a one-line comment naming the invariant.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by KSXGitHub — COMMENTED** [review]
Self-review (iteration 2) on commit `cc82787`. Six new findings, all minor / nit. Plan: fix four (source-link prefix, root-collision guard, error-message hint, missing test) and add explanatory comments for the two unreachable edge cases (case-insensitive FS, index single-write invariant). No critical or major issues.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tools/gen-docs/src/main.rs [hunk 16]
```diff
@@ -131,3 +180,113 @@ fn main() -> ExitCode {
...
+/// Compute the relative-path prefix the markdown renderer needs
+/// to climb from a file under `rules_dir` back up to the repo
+/// `root`. Returned with a trailing `/`, so callers can append a
+/// path under `root` directly (e.g. `"../" + "src/rules/foo.rs"`).
+///
+/// Returns `Err` when `rules_dir` is the same directory as `root`
+/// (the catalogue's `README.md` would collide with the project's
+/// own) or when it resolves to a path outside `root` (the renderer
+/// has no way to produce a correct relative link in that case).
+fn source_link_prefix_for(rules_dir: &Path, root: &Path) -> Result<String, String> {
```
Comment (KSXGitHub) [inline tools/gen-docs/src/main.rs:197]: **Finding 1 (major): `source_link_prefix_for` miscounts depth on `..` in input**

`std::path::absolute` only joins relatives onto cwd; it does *not* normalise `..` or `.` components. So `--rules-dir rules/../rules` yields an absolute path containing `rules/../rules`, `strip_prefix` returns `rules/../rules` (3 components), and `depth = 3` produces `"../../../"` — broken `Source:` links three levels too deep. `rules/sub/..` (semantically `rules/`) has the same shape and produces depth 3.

**Assessment:** Valid. **Will fix:** walk `rel.components()` and count only `Component::Normal`; reject `ParentDir` (`..`) explicitly so the user gets an error instead of broken links.

---
_Generated by [Claude Code](https://claude.ai/code)_

Comment (KSXGitHub) [inline tools/gen-docs/src/main.rs:197]: **Finding 2 (major): no direct tests for `source_link_prefix_for`**

The most subtle new logic in iteration 3 (path absolutisation, depth-counting, root-collision and outside-root rejection) has zero direct tests. The `check_rules_dir` / `write_rules_dir` tests pass a literal `"../"` and never exercise the helper.

**Assessment:** Valid coverage gap. **Will fix:** add a `tests` module in `main.rs` covering: depth 1, depth 3, `rules_dir == root` → Err, `rules_dir` outside root → Err, `..` in input → Err (validates Finding 1's fix).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tools/gen-docs/src/check_md.rs [hunk 17]
```diff
@@ -0,0 +1,363 @@
...
+        if needs_write {
+            fs::write(&path, body)
+                .unwrap_or_else(|error| panic!("failed to write {}: {error}", path.display()));
+        }
+        if file_name == INDEX_FILE_NAME {
+            // Plain assignment, not `|=`: `BTreeMap` keys are
+            // unique, so the loop visits `INDEX_FILE_NAME` at most
+            // once. If a future change ever adds a second managed
+            // index file, replace this with the explicit OR.
+            summary.index_changed = needs_write;
```
Comment (KSXGitHub) [inline tools/gen-docs/src/check_md.rs]: **Finding 3 (nit): `index_changed = needs_write` should be `|=` defensively**

The single-visit invariant the comment relies on (BTreeMap dedupes by key) is sound today, but the comment itself flags a future retry / multi-index refactor as a hazard. `|=` is the same behaviour under the invariant and stays correct if it's ever violated — zero cost to harden now.

**Assessment:** Valid. **Will fix:** swap to `|=`, drop the if-the-future hedge from the comment.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tools/gen-docs/src/check_md.rs [hunk 18]
```diff
@@ -0,0 +1,363 @@
...
+                    "internal error: two rules rendered to the same filename `{}`",
+                    rule_file_name(rule),
+                );
+            }
+        }
+        files.insert(INDEX_FILE_NAME.to_owned(), render_index_md(rules));
+        ExpectedTree { files }
+    }
+
+    fn is_managed(&self, file_name: &str) -> bool {
```
Comment (KSXGitHub) [inline tools/gen-docs/src/check_md.rs:56]: **Finding 4 (minor): case-insensitive FS data-swap is worse than the comment suggests**

On macOS APFS (case-insensitive by default), a user-created `readme.md` and the catalogue's `README.md` are the *same inode*. `read_dir` returns whichever case is on disk; `is_managed` returns false; the orphan-delete then removes the file the renderer was about to write — and the subsequent write succeeds, but the user-typed contents the file used to hold are gone. Not just "deleted as orphan" — *silently swapped*.

**Assessment:** Valid; the comment understates the failure mode. **Will fix:** add a case-insensitive guard in the orphan-delete loop. If an on-disk name case-insensitively matches any expected name, skip the deletion. On case-sensitive filesystems (Linux/typical CI), the only files our tool ever writes are the canonical casings, so this guard never fires; on case-insensitive filesystems it prevents the swap.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by KSXGitHub — COMMENTED** [review]
Self-review (iteration 3) on commit `3644561`. Four findings: two majors I'll fix (depth miscount on `..` input; missing direct tests for `source_link_prefix_for`), one nit (`|=` for defensive index-change tracking), one minor (case-insensitive FS data-swap hazard worse than commented — adding an orphan-delete guard). I expect this iteration to be closer to terminal; finding-1's fix removes a real footgun, finding-2 closes the coverage gap exposing it.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by KSXGitHub — COMMENTED** [review]
Self-review iteration 4 on `885e486` — one finding, fixed in this push.

**Heading-level cascade (medium):** rule title rendered as h1 and `## Configuration` as h2, but `rule.doc_markdown` was injected verbatim, so source-level `### What it does` lines landed at h3 — the same TOC depth as the h3 config field headings, so a reader couldn't distinguish "prose section about the lint" from "config field of the lint". The fix promotes every ATX h2-h6 heading in the doc-comment block up one level when rendering, with a state machine that skips fenced and indented code blocks so example output containing `###` lines isn't silently rewritten. New tests cover promotion, the h1-preservation edge, code-block transparency, and the 7+-hash non-heading case.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by KSXGitHub — COMMENTED** [review]
Self-review iteration 5 on `8c46a75` — one finding, fixed in this push.

**Fence-marker / length tracking in `promote_headings` (minor, latent):** the previous boolean toggle treated any `` ``` `` or `~~~` line as a fence delimiter regardless of which character opened the current fence — violating CommonMark §4.5 which requires a close to match the opener's character and be at least as long. Current rule sources use only 3-backtick fences, so the bug never fired in practice, but the doc-comment claimed transparency for fenced blocks. Replaced the `bool` with `Option<(u8, usize)>` and added tests for the mixed-marker (`` ``` `` outer / `~~~` body) and length-mismatch (`` ```` `` outer / `` ``` `` body) cases.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by KSXGitHub — COMMENTED** [review]
Self-review iteration 6 on `8c46a75` — **no new findings; loop terminating.**

The surface is clean across `main.rs`, `check_md.rs`, and `render_md.rs`. Iterations 1-5 covered: typography + index format (em-dash → colon, multi-line bullet list), generated-file banner, UTF-8 filename guard, tempdir counter, write-summary clarity, source-link prefix helper with `..`/outside-root/root-collision rejection and direct tests, defensive `index_changed |=`, case-insensitive orphan-delete guard, drift-message canonicalisation, heading promotion with fence transparency, and CommonMark-spec-compliant fence tracking by marker character and length. 39 unit tests pass; the in-tree `rules/` round-trips cleanly through `check-md`; the helper `source_link_prefix_for` has six direct tests covering depth-1, depth-3, root-collision, outside-root, `..` rejection, and `.` transparency.

Out-of-scope items remain out of scope: non-atomic writes and panic-on-IO match the existing `run_html` style and would be a separate refactor; `planned-rules/` cross-reference cleanup the maintainer flagged earlier remains a follow-up PR.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tools/gen-docs/src/main.rs [hunk 19]
```diff
@@ -131,3 +180,206 @@ fn main() -> ExitCode {
...
+        )
+    })?;
+    let rel = abs_rules.strip_prefix(&abs_root).map_err(|_| {
+        format!(
+            "rules_dir `{}` is not inside --root `{}`; the markdown renderer can only \
+             build relative `Source:` links when rules_dir is under the repo root",
+            rules_dir.display(),
+            root.display(),
+        )
+    })?;
```
[the remark this answers is not in the document]
Comment (KSXGitHub) [reply tools/gen-docs/src/main.rs:236]: Empirically `std::path::absolute` *does* normalise `.` segments on the platforms std supports — `std::path::absolute("/repo/.")` returns `/repo`, not `/repo/.`, so the `strip_prefix` failure scenario doesn't occur. Pinned the behaviour with a regression test (`source_link_prefix_for_trailing_dot_on_root` in `64fa7e5`) so a future loosening of the std contract would surface immediately. `..` segments are a separate concern and are still explicitly rejected.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tools/gen-docs/src/render_md.rs [hunk 20]
```diff
@@ -0,0 +1,575 @@
...
+                }
+                Some(_) => {
+                    // Mismatched marker or shorter run: this is
+                    // body content of the open fence, not a close
+                    // marker. Stay inside the fence.
+                }
+            }
+            out.push_str(line);
+            continue;
+        }
```
[the remark this answers is not in the document]
Comment (KSXGitHub) [reply tools/gen-docs/src/render_md.rs]: Valid — fixed in `64fa7e5`. The close-fence check now requires the post-marker tail to be blank per CommonMark §4.5, so a nested `` ```rust `` line inside a wider outer fence stays in body context. Added `promote_headings_rejects_info_string_as_close_fence` covering the 4-backtick-outer / 3-backtick-info-stringed-inner case.

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (KSXGitHub) [reply tools/gen-docs/src/render_md.rs]: Valid — fixed in `2888a11`. `promote_headings` now computes `leading_spaces` first and only treats the line as a fence marker when it's within CommonMark §4.5's 0-3-space allowance; 4+ leading spaces fall through to the indented-code-block branch. Added `promote_headings_treats_indented_fence_marker_as_indented_code` covering the `    ```` ` opener case.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tools/gen-docs/src/render_md.rs [hunk 21]
```diff
@@ -0,0 +1,610 @@
...
+        if open_fence.is_some() {
+            out.push_str(line);
+            continue;
+        }
+        let leading_spaces = line.bytes().take_while(|&b| b == b' ').count();
```
[the remark this answers is not in the document]
Comment (KSXGitHub) [reply tools/gen-docs/src/render_md.rs]: Valid — fixed in `2888a11`. `promote_headings` now computes `leading_spaces` first and only treats the line as a fence marker when it's within CommonMark §4.5's 0-3-space allowance; 4+ leading spaces fall through to the indented-code-block branch. Regression test `promote_headings_treats_indented_fence_marker_as_indented_code` covers the `    ```` `-opener case.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: tools/gen-docs/src/main.rs [hunk 22]
```diff
@@ -131,3 +180,218 @@ fn main() -> ExitCode {
...
```
[the remark this answers is not in the document]
Comment (KSXGitHub) [reply tools/gen-docs/src/main.rs:236]: Valid — fixed in `2888a11`. Added a `fs::symlink_metadata` guard in `source_link_prefix_for` that refuses any `rules_dir` whose leaf is itself a symbolic link, before the lexical `strip_prefix` check (which can't see through symlinks). Catches the most common footgun — `rules -> /etc` then `write-md ./rules` — without the canonicalize-vs-create-vs-check-md ordering complexity. Regression test `source_link_prefix_for_rejects_symlink_rules_dir` covers the case on Unix; skipped on Windows where unprivileged symlink creation isn't reliable.

A symlink at a *parent* path component (e.g. the cwd resolving through a symlink) still slips past; that's out of scope — the user controls their checkout and a symlinked-repo-root is a legitimate setup.

---
_Generated by [Claude Code](https://claude.ai/code)_
